File: src/main/java/june_2025/MyCalendar.java
package june_2025;

public class MyCalendar {

    Node root;
    public MyCalendar() {
        root = null;
    }

    public boolean book(int startTime, int endTime) {
        if (root == null) {
            root = new Node(startTime, endTime);
            return true;
        }
        return insert(root, startTime, endTime);
    }

    private boolean insert(Node node, int start, int end) {
        if (end <= node.startTime) {
            if (node.left == null) {
                node.left = new Node(start, end);
                return true;
            }
            return insert(node.left, start, end);
        } else if (start >= node.endTime) {
            if (node.right == null) {
                node.right = new Node(start, end);
                return true;
            }
            return insert(node.right, start, end);
        }
        return false;
    }

    static class Node {
        int startTime;
        int endTime;
        Node left;
        Node right;

        public Node(int startTime, int endTime) {
            this.startTime = startTime;
            this.endTime = endTime;
        }
    }

}

File: python/second/my_calendar.py
class MyCalendar(object):

    def __init__(self):
        self.root = None        

    def book(self, startTime, endTime):
        """
        :type startTime: int
        :type endTime: int
        :rtype: bool
        """
        if not self.root:
            self.root = Node(startTime, endTime)
            return True
        
        return self.insert(self.root, startTime, endTime)
    
    def insert(self, node, start, end):
        if end <= node.start:
            if not node.left:
                node.left = Node(start, end)
                return True
            return self.insert(node.left, start, end)
        elif start >= node.end:
            if not node.right:
                node.right = Node(start, end)
                return True
            return self.insert(node.right, start, end)
        else:
            return False


class Node(object):
    def __init__(self, start, end, left = None, right = None):
        self.start = start
        self.end = end
        self.left = left
        self.right = right
